test/controllers/tasks_controller_test.rb

```diff
@@ -3,124 +3,191 @@
+      
     end
-
+    
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+      
+      get edit_task_path(-1)
+      
+      must_respond_with :redirect 
+      must_redirect_to tasks_path
     end
   end
-
+  
   # Uncomment and complete these tests for Wave 3
   describe "update" do
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
     it "can update an existing task" do
-      # Your code here
+      
+      #Arrange:
+      #step one get an existing book
+      #Step two once you have the existing book, change the name of the task. 
+      existing_task = task
+      updated_task_form_data = {
+        task: {
+          name: "cs fun",
+          description: "Recursion practice",
+          completed: Date.today
+        }
+      }
+      #Act:
+      #actually update the book data with the new hash form data you created.
+      patch task_path(existing_task.id), params: updated_task_form_data
+      
+      # Assert
+      #Step Three once you change the fields, check to see if they are different than original 
+      #expect(Book.find_by(id: existing_book.id).title).must_equal "Name that you changed book to"
+      
+      expect(Task.find_by(id: existing_task.id).name).must_equal "cs fun"
+      expect(Task.find_by(id: existing_task.id).description).must_equal "Recursion practice"
+      expect(Task.find_by(id: existing_task.id).completed).must_equal Date.today
     end
-
+    
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      # Arrange:
+      
+      updated_task_form_data = {
+        task: {
+          name: "cs fun",
+          description: "Recursion practice",
+          completed: Date.today
+        }
+      }
+      
+      patch task_path(-1), params: updated_task_form_data
+      # Assert:
+      must_respond_with :redirect
+      must_redirect_to root_path
     end
   end
-
+  
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
-
+    it "successfully deletes an existing task and then redirects to home page" do
+      created_task = task
+      expect{
+        delete task_path(created_task.id)
+      }.must_differ "Task.count", -1
+      must_redirect_to root_path 
+    end
   end
-
+  
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    
+    it "successfully marks a task as complete with completed date" do
+      #Arrange
+      #when the user clicks on the complete button, the program will go back into the data base and
+      #change the completed date from nil and set it to the current date.
+      
+      existing_task =  Task.create(name: "cs fun", description: "recursion practice", completed: nil)
+      patch complete_path(existing_task)
+      expect(Task.find_by(id: existing_task.id).completed).must_be_instance_of  ActiveSupport::TimeWithZone
```

Comment:
This test would be perfect and complete with an assertion to check the response, like `must_redirect_to` or something